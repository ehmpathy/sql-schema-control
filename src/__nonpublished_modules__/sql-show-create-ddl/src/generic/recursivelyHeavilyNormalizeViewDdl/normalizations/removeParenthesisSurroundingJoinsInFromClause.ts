/**
 * strip out the first and last paren of the "FROM" section, IF it is defined;
 *
 * show create likes to wrap the whole "FROM" section w/ parens... its unreasonable to ask people to do that too.
 */
export const removeParenthesisSurroundingJoinsInFromClause = ({
  flattenedSql,
}: {
  flattenedSql: string;
}) => {
  // check that this sql has this situation going on, if not, do nothing
  const ddlHasParenOpenRightAfterFromClause =
    !!flattenedSql.match(/\sfrom\s+\(/gi); // note, we're not worried about subqueries because we expect flattenedSql as input
  if (!ddlHasParenOpenRightAfterFromClause) return flattenedSql; // do nothing if the situation does not exist

  // split DDL on the from statement, since we know it exists
  const flattenedSqlParts = flattenedSql.split(/\sfrom\s/gi);
  if (flattenedSqlParts.length !== 2)
    throw new Error(
      'not exactly two parts after splitting on from; unexpected',
    ); // fail fast

  // remove the parens that encapsulate each join
  const fromClauseWithoutParens = flattenedSqlParts[1]
    .replace(/\(/g, '')
    .replace(/\)/g, ''); // note: we replace all parens, since subqueries are taken care of

  // join them back and the ddl w/o this mess
  const flattenedSqlWithoutOpenCloseParenInFromClause =
    flattenedSqlParts[0] + '\nfrom\n' + fromClauseWithoutParens; // tslint:disable-line prefer-template

  // return without the parens
  return flattenedSqlWithoutOpenCloseParenInFromClause;
};
