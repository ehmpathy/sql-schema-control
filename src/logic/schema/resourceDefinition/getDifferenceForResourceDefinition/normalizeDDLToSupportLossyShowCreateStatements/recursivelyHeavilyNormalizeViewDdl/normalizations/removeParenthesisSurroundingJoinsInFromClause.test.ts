import { removeParenthesisSurroundingJoinsInFromClause } from './removeParenthesisSurroundingJoinsInFromClause';

describe('removeParenthesisSurroundingJoinsInFromClause', () => {
  it('should remove them from every join', () => {
    const exampleSql = `
select
  s.id as id,
  s.uuid as uuid,
  s.name as name,
  s.created_at as created_at,
  v.effective_at as effective_at,
  v.created_at as updated_at
from
  (
    (
      contractor s
      join contractor_cvp cvp on s.id = cvp.contractor_id
    )
    join contractor_version v on v.id = cvp.contractor_version_id
  )
    `.trim();
    const normalizedSql = removeParenthesisSurroundingJoinsInFromClause({ flattenedSql: exampleSql });
    expect(normalizedSql).not.toContain('(');
    expect(normalizedSql).not.toContain(')');
    expect(normalizedSql).toMatchSnapshot();
  });
});
