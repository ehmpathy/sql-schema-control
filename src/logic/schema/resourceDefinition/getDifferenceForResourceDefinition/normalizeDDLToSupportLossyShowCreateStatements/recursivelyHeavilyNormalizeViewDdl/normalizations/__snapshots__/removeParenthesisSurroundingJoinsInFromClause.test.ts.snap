// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removeParenthesisSurroundingJoinsInFromClause should remove them from every join 1`] = `
"select
  s.id as id,
  s.uuid as uuid,
  s.name as name,
  s.created_at as created_at,
  v.effective_at as effective_at,
  v.created_at as updated_at
from
  
    
      contractor s
      join contractor_cvp cvp on s.id = cvp.contractor_id
    
    join contractor_version v on v.id = cvp.contractor_version_id
  "
`;
